### PR TITLE
net/upnp: enable STUN and allow LAN subnet override

### PR DESCRIPTION
**Pull request by Tawmu**
- Fixes issue #2974 by allowing override of a single LAN interface subnet
- Allows configuration of STUN server for external IP prediction

### CODE REVIEW

**Comment by Tawmu** [comment]
@fichtner any way we can progress this?

File: net/upnp/src/etc/inc/plugins.inc.d/miniupnpd.inc [hunk 1]
```diff
@@ -150,7 +150,11 @@ function miniupnpd_configure_do($verbose = false)
             if ($if != $iface) {
                 $addr = get_interface_ip($iface);
                 if (!empty($addr)) {
-                    $config_text .= "listening_ip={$if}\n";
+                    if(isset($upnp_config['overridesubnet'])) {
+                        $config_text .= "listening_ip={$if}{$upnp_config['overridesubnet']}\n";
```
Comment (fichtner) [inline net/upnp/src/etc/inc/plugins.inc.d/miniupnpd.inc:154]: what sort of data is implied here? there's no separation between first and second variable

File: net/upnp/Makefile [hunk 2]
```diff
@@ -1,5 +1,5 @@
 PLUGIN_NAME=		upnp
-PLUGIN_VERSION=		1.4
+PLUGIN_VERSION=		1.5
 PLUGIN_REVISION=	2
```
Comment (fichtner) [inline net/upnp/Makefile:3]:
```suggestion
PLUGIN_VERSION=		1.5
```

File: net/upnp/src/etc/inc/plugins.inc.d/miniupnpd.inc [hunk 3]
```diff
@@ -167,7 +171,7 @@ function miniupnpd_configure_do($verbose = false)
 
         if (!empty($ifaces_active)) {
             /* override wan ip address, common for carp, etc */
-            if (!empty($upnp_config['overridewanip'])) {
+            if (!empty($upnp_config['overridewanip']) && empty($upnp_config['stun_host'])) {
```
Comment (fichtner) [inline net/upnp/src/etc/inc/plugins.inc.d/miniupnpd.inc:174]: if this change is needed the logic should continue with `stun_host` configuration instead of skipping unrelated lines to below.

File: net/upnp/src/etc/inc/plugins.inc.d/miniupnpd.inc [hunk 4]
```diff
@@ -226,6 +230,16 @@ function miniupnpd_configure_do($verbose = false)
             $config_text .= "min_lifetime=120\n";
             $config_text .= "max_lifetime=86400\n";
 
+            /* configure STUN server if needed */
+            if (!empty($upnp_config['stun_host']) && empty($upnp_config['overridewanip'])) {
+                if(empty($upnp_config['stun_port'])) {
+                    $upnp_config['stun_port'] == '3478';
```
Comment (fichtner) [inline net/upnp/src/etc/inc/plugins.inc.d/miniupnpd.inc:236]: wouldn't do this if you use `??` underneath... which is safer than tainting configuration data (it could end up being saved)

File: net/upnp/src/www/services_upnp.php [hunk 5]
```diff
@@ -104,6 +105,12 @@ function miniupnpd_validate_port($port)
     if (!empty($pconfig['overridewanip']) && !is_ipaddr($pconfig['overridewanip'])) {
         $input_errors[] = gettext('You must specify a valid ip address in the \'Override WAN address\' field');
     }
+    if(!empty($pconfig['overridewanip']) && !empty($pconfig['stun_host'])) {
```
Comment (fichtner) [inline net/upnp/src/www/services_upnp.php:108]: should have some sort of data validation...

File: net/upnp/src/www/services_upnp.php [hunk 6]
```diff
@@ -104,6 +105,12 @@ function miniupnpd_validate_port($port)
...
+        $input_errors[] = gettext('You cannot override the WAN IP if you have a STUN host set');
+    }
+    if(!empty($pconfig['overridesubnet']) && count($pconfig['iface_array']) > 1) {
```
Comment (fichtner) [inline net/upnp/src/www/services_upnp.php:111]: data validation missing here as well

**Comment by fichtner** [comment]
Thanks so far! As a side node it would be better to split STUN and override IP for clearer PR audit and faster inclusion.

File: net/upnp/Makefile [hunk 7]
```diff
@@ -1,5 +1,5 @@
...
 PLUGIN_DEPENDS=		miniupnpd
```
Comment (fichtner) [inline net/upnp/Makefile:4]:
```suggestion
PLUGIN_VERSION=		1.5
PLUGIN_DEPENDS=		miniupnpd-devel
```

File: net/upnp/src/etc/inc/plugins.inc.d/miniupnpd.inc [hunk 8]
```diff
@@ -150,7 +150,11 @@ function miniupnpd_configure_do($verbose = false)
             if ($if != $iface) {
                 $addr = get_interface_ip($iface);
                 if (!empty($addr)) {
-                    $config_text .= "listening_ip={$if}\n";
+                    if(isset($upnp_config['overridesubnet'])) {
```
Comment (fichtner) [inline net/upnp/src/etc/inc/plugins.inc.d/miniupnpd.inc:153]:
```suggestion
                    if (isset($upnp_config['overridesubnet'])) {
```

File: net/upnp/src/etc/inc/plugins.inc.d/miniupnpd.inc (hunk 1 above)
Comment (Tawmu) [reply net/upnp/src/etc/inc/plugins.inc.d/miniupnpd.inc:151]: Didn't mean to start a formal review, leaving comment here to explain this one.

This is to allow the user to override the subnet as defined by opnsense so that miniupnpd can respond to upnp requests. Useful if running miniupnpd across VLANs for example.

see: https://github.com/miniupnp/miniupnp/issues/330#issuecomment-434781573

**Comment by Tawmu** [comment]
Hopefully this is better - have taken points above and amended as necessary with clarification on the listening_ip directive.

File: net/upnp/src/www/services_upnp.php [hunk 9]
```diff
@@ -255,6 +268,44 @@ function miniupnpd_validate_port($port)
...
                     </tr>
+
+                    <tr>
+                        <td><a id="help_for_override_subnet" href="#" class="showhelp"><i class="fa fa-info-circle"></i></a> <?=gettext("Interface subnet override");?></td>
+                        <td>
+                            <select name="overridesubnet" class="selectpicker" id="overridesubnet">
+                                <option value="" <?=!empty($pconfig['overridesubnet']) ? "selected=\"selected\"" : "";?>>Interface default</option>
+<?php
+                                for ($i = 32; $i >= 1; $i--) { ?>
+                                    <option value="/<?=$i;?>" <?=!empty($pconfig['overridesubnet']) && $pconfig['overridesubnet'] == "/".$i ? "selected=\"selected\"" : "";?>>/<?=$i;?></option>
```
Comment (fichtner) [inline net/upnp/src/www/services_upnp.php:279]: it's probably safer to validate for subnet mask and add the "/" in the configuration. /32 is basically the default like when we omit the override?

**Comment by tiurs** [comment]
Is there a way I can apply this fix now or do I have to wait till it is released?

**Comment by fichtner** [comment]
going to merge to work on this... thanks!

**Comment by fichtner** [comment]
@Tawmu make sure to test 7144429a9
